### PR TITLE
Recoverability custom error queue

### PR DESCRIPTION
**Pull request by danielmarbach**
Connects to https://github.com/Particular/V6Launch/issues/69

Based on top of #3939 

# Content

Introduces the possibility to specify a custom error queue inside the custom recoverability policy. The standard queue is provided on the RecoverabilityConfiguration. Implementors of a custom policy a free to ignore the parameter.

## Questions
* Should the notifications for move to error contain the error queue address? I think they should
* Should we move `SendFailedMessagesTo` under the umbrella of the new recoverability configuration API provided by #3939 ? I think we should because it aligns those concepts. of course this would be an effort to migrate all samples but that shouldn't hinder us. So instead

```
configuration.SendFailedMessagesTo("errorQueue");
```

you'd do

```
configuration.Recoverability().Failed(failed =>failed.Queue("errorQueue"));
```

### CODE REVIEW

File: src/NServiceBus.Core.Tests/Recoverability/DefaultRecoverabilityPolicyTests.cs [hunk 1]
```diff
@@ -3,41 +3,54 @@
...
-            var recoverabilityAction = policy.Invoke(errorContext);
+            var recoverabilityAction = policy(errorContext);
 
             Assert.IsInstanceOf<ImmediateRetry>(recoverabilityAction, "We should have one immediate retry left. It is second delivery attempt and we configured immediate reties to 2.");
         }
 
         [Test]
-        public void When_max_immediate_retries_exceeded_should_return_delayed_retry()
+        [TestCase(0)]
+        [TestCase(1)]
```
Comment (timbussmann) [inline src/NServiceBus.Core.Tests/Recoverability/DefaultRecoverabilityPolicyTests.cs]: what's the reason for the testcases here? they don't even seem to test edge cases like 2/3?

File: src/NServiceBus.Core.Tests/Recoverability/MoveToErrorsExecutorTests.cs [hunk 2]
```diff
@@ -18,23 +18,25 @@ public void Setup()
...
             dispatcher = new FakeDispatcher();
             staticFaultMetadata = new Dictionary<string, string>();
-            moveToErrorsExecutor = new MoveToErrorsExecutor(dispatcher, ErrorQueueAddress, staticFaultMetadata);
+            moveToErrorsExecutor = new MoveToErrorsExecutor(dispatcher, staticFaultMetadata, headers => { });
         }
 
         [Test]
-        public async Task MoveToErrorQueue_should_dispatch_message_to_error_queue()
+        [TestCase(ErrorQueueAddress)]
+        [TestCase("random_error_queue")]
```
Comment (timbussmann) [inline src/NServiceBus.Core.Tests/Recoverability/MoveToErrorsExecutorTests.cs]: do we need  testcases here? At first glance `"random_error_queue"` should be good enough to verify the usage of the passed parameter?

File: src/NServiceBus.Core/Recoverability/DefaultRecoverabilityPolicy.cs [hunk 3]
```diff
@@ -4,58 +4,102 @@ namespace NServiceBus
...
+            {
+                return RecoverabilityAction.MoveToError(config.Failed.ErrorQueue);
+            }
+
+            if (errorContext.NumberOfFailedImmediateDeliveryAttempts <= config.Immediate.MaxNumberOfRetries)
+            {
+                return RecoverabilityAction.ImmediateRetry();
+            }
+
+            Logger.InfoFormat("Giving up First Level Retries for message '{0}'.", errorContext.Message.MessageId);
```
Comment (timbussmann) [inline src/NServiceBus.Core/Recoverability/DefaultRecoverabilityPolicy.cs]: we could wrap that in an `if(config.Immediate.MaxNumberOfRetries > 0)` but not sure it would add much value.

File: src/NServiceBus.Core/Recoverability/DelayedRetryExecutor.cs [hunk 4]
```diff
@@ -20,10 +20,10 @@ public DelayedRetryExecutor(string endpointInputQueue, IDispatchMessages dispatc
 
         public async Task<int> Retry(IncomingMessage message, TimeSpan delay, TransportTransaction transportTransaction)
         {
-            var currentDelayedRetriesAttempt = message.GetCurrentDelayedRetries() + 1;
-
             var outgoingMessage = new OutgoingMessage(message.MessageId, new Dictionary<string, string>(message.Headers), message.Body);
 
+            var currentDelayedRetriesAttempt = message.GetCurrentDelayedRetries() + 1;
```
Comment (timbussmann) [inline src/NServiceBus.Core/Recoverability/DelayedRetryExecutor.cs]: any particular reason for this move?

File: src/NServiceBus.Core/Recoverability/Recoverability.cs [hunk 5]
```diff
@@ -25,10 +25,10 @@ public Recoverability()
                 "Message recoverability is only relevant for endpoints receiving messages.");
             Defaults(settings =>
             {
-                settings.SetDefault(SlrNumberOfRetries, DefaultSecondLevelRetryPolicy.DefaultNumberOfRetries);
-                settings.SetDefault(SlrTimeIncrease, DefaultSecondLevelRetryPolicy.DefaultTimeIncrease);
-
+                settings.SetDefault(SlrNumberOfRetries, DefaultNumberOfRetries);
+                settings.SetDefault(SlrTimeIncrease, DefaultTimeIncrease);
```
Comment (timbussmann) [inline src/NServiceBus.Core/Recoverability/Recoverability.cs]: where does the error queue get set?

Comment (timbussmann) [inline src/NServiceBus.Core/Recoverability/Recoverability.cs]: nevermind, saw it. Can't we set a default here and override it in the `MoveFailedBlabla` (which we then can remove anyways?)

File: src/NServiceBus.Core/Recoverability/RecoverabilitySettings.cs [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+
+            return this;
+        }
+
+        /// <summary>
+        /// Configures the delay interval increase for each failed second level retry attempt.
+        /// </summary>
+        public DelayedRetriesSettings TimeIncrease(TimeSpan timeIncrease)
+        {
+            Guard.AgainstNegative(nameof(timeIncrease), timeIncrease);
```
Comment (timbussmann) [inline src/NServiceBus.Core/Recoverability/RecoverabilitySettings.cs]: didn't simon add some more advanced checks in the old code to make sure that it's within a reasonable timespan?

**Comment by timbussmann** [comment]
> Should the notifications for move to error contain the error queue address?

sounds reasonable :+1: 

> Should we move SendFailedMessagesTo under the umbrella of the new recoverability configuration API provided

yes, please! :+1: :tada: (while adding a default value)

File: src/NServiceBus.Core.Tests/Recoverability/DefaultRecoverabilityPolicyTests.cs (hunk 1 above)
Comment (tmasternak) [inline src/NServiceBus.Core.Tests/Recoverability/DefaultRecoverabilityPolicyTests.cs]: Agree. I've changed that to fixed value 2 which is the smallest one that should exceed immediate retry count attempts.

File: src/NServiceBus.Core.Tests/Recoverability/MoveToErrorsExecutorTests.cs (hunk 2 above)
Comment (tmasternak) [inline src/NServiceBus.Core.Tests/Recoverability/MoveToErrorsExecutorTests.cs]: Agree. Switched to single `random_error_queue` as a value we test against

File: src/NServiceBus.Core/Recoverability/DefaultRecoverabilityPolicy.cs (hunk 3 above)
Comment (tmasternak) [inline src/NServiceBus.Core/Recoverability/DefaultRecoverabilityPolicy.cs]: I would say we should remove it. I'm not sure what value if brings and the proper logging is done in the `RecoverabilityExecutor` for example for [slr](https://github.com/Particular/NServiceBus/blob/cdb383223bbf54ed1c5f006ef2fc68ce6ad9150b/src/NServiceBus.Core/Recoverability/RecoverabilityExecutor.cs#L66)

File: src/NServiceBus.Core/Recoverability/DelayedRetryExecutor.cs (hunk 4 above)
Comment (tmasternak) [inline src/NServiceBus.Core/Recoverability/DelayedRetryExecutor.cs]: No only to move declaration closer to actual usage

File: src/NServiceBus.Core/Recoverability/Recoverability.cs (hunk 5 above)
Comment (tmasternak) [inline src/NServiceBus.Core/Recoverability/Recoverability.cs]: @timbussmann I'm not sure I'm following :P

File: src/NServiceBus.Core/Recoverability/RecoverabilitySettings.cs (hunk 6 above)
Comment (tmasternak) [inline src/NServiceBus.Core/Recoverability/RecoverabilitySettings.cs]: I could not find anything both in v5 and previous v6.

**Comment by tmasternak** [comment]
@tim: can split off moving `SendFailedMessagesTo` to a separate PR? It probably deserves pinging relevant people to make sure we did not miss anything.

**Comment by timbussmann** [comment]
@tmasternak good with that

File: src/NServiceBus.Core/Recoverability/RecoverabilitySettings.cs (hunk 6 above)
Comment (timbussmann) [inline src/NServiceBus.Core/Recoverability/RecoverabilitySettings.cs]: ah yeah that was a different API, nevermind

**Comment by danielmarbach** [comment]
pulled in the description of the previous (now closed) PR
